godot-macros/src/docs.rs

```diff
@@ -0,0 +1,179 @@
+                        doc.into_iter()
+                            .map(|x| match x {
+                                // venial??
+                                proc_macro2::TokenTree::Literal(x) => x
+                                    .to_string()
+                                    .trim_start_matches('"')
+                                    .trim_end_matches('"')
+                                    .to_string(),
+                                _ => unreachable!(),
+                            })
+                            .reduce(|a, b| format!("{a} {b}"))
+                            .unwrap(),
+                    ),
+                    AttributeValue::Group(_, _) => None,
+                    // valid forms of #[doc] are
+                    // #[doc(hidden|inline|...)] or
+                    // #[doc = string]
+                    _ => unreachable!(),
+                }
+            })?
+        })
+        .collect::<String>();
+    md2bb::convert(&md)
+}
+
+fn signal(signal: SignalDefinition) -> String {
+    let name = signal.signature.name;
+    let params = params(signal.signature.params.iter().filter_map(|(x, _)| match x {
+        FnParam::Receiver(_) => None,
+        FnParam::Typed(y) => Some((y.name.clone(), y.ty.clone())),
+    }));
+    let desc = docs(signal.external_attributes);
+    format!(
+        r#"
+<signal name="{name}">
+  {params}
+  <description>
+  {desc}
+  </description>
+</signal>
+"#
+    )
+}
+
+fn constant(cnst: Constant) -> String {
+    let name = cnst.name.to_string();
+    let value = cnst
+        .initializer
+        .map(|x| x.to_token_stream().to_string())
+        .unwrap_or("null".into());
+    let docs = docs(cnst.attributes);
+    format!(r#"<constant name="{name}" value="{value}">{docs}</constant>"#)
+}
+
+pub fn member(member: Field) -> String {
+    let name = member.name.to_string();
+    let ty = member.ty.to_token_stream().to_string();
+    let docs = docs(member.attributes);
+    let default = member.default.to_token_stream().to_string();
+    format!(r#"<member name="{name}" type="{ty}" default="{default}">{docs}</member>"#)
+}
+
+fn params(params: impl Iterator<Item = (Ident, TypeExpr)>) -> String {
+    params
+        .zip(0..)
+        .fold(String::default(), |mut acc, ((name, ty), index)| {
+            use std::fmt::Write;
+            _ = write!(
+                acc,
+                r#"<param index="{index}" name="{name}" type="{ty}" />"#,
+                ty = ty.to_token_stream()
+            );
+            acc
+        })
+}
+
+pub fn imethod(method: Function) -> String {
+    let name = method.name.to_string();
+    let ret = method
+        .return_ty
+        .map(|x| x.to_token_stream().to_string())
+        .unwrap_or("void".into());
+    let params = params(method.params.iter().filter_map(|(x, _)| match x {
+        FnParam::Receiver(_) => None,
+        FnParam::Typed(y) => Some((y.name.clone(), y.ty.clone())),
+    }));
```

Comment:
i think `params` (the function) should do this filtering internally, rather than needing to filter before passing it in

Reply:
Cant do that for `method` as it has a different type of argument.